In [2]:
%pip install pandas
%pip install numpy
%pip install matplotlib 
%pip install seaborn    
%pip install scikit-learn   


Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [3]:
import os 
import zipfile
import pandas as pd

zip_file = "/Users/anovayoungers/Downloads/TheMovieDataset.zip"

extract_folder = "/Users/anovayoungers/Downloads/TheMovieDataset"

# 1. Unzip the file into the Capstone folder
with zipfile.ZipFile(zip_file, 'r') as zip_ref:
    zip_ref.extractall(extract_folder)
    
# 2. Load each CSV into a DataFrame
credits_df         = pd.read_csv(os.path.join(extract_folder, "credits.csv"))
keywords_df        = pd.read_csv(os.path.join(extract_folder, "keywords.csv"))
links_small_df     = pd.read_csv(os.path.join(extract_folder, "links_small.csv"))
links_df           = pd.read_csv(os.path.join(extract_folder, "links.csv"))
movies_metadata_df = pd.read_csv(os.path.join(extract_folder, "movies_metadata.csv"))
ratings_small_df   = pd.read_csv(os.path.join(extract_folder, "ratings_small.csv"))
ratings_df         = pd.read_csv(os.path.join(extract_folder, "ratings.csv"))

# 3. Inspect the loaded DataFrames
print("credits_df shape:", credits_df.shape)
print("keywords_df shape:", keywords_df.shape)
print("links_small_df shape:", links_small_df.shape)
print("links_df shape:", links_df.shape)
print("movies_metadata_df shape:", movies_metadata_df.shape)
print("ratings_small_df shape:", ratings_small_df.shape)
print("ratings_df shape:", ratings_df.shape)

# Peek at the first few rows of one DataFrame (as an example)
print(movies_metadata_df.head())

/var/folders/3y/w5ts8qfs3sl47n2k1dprbgs00000gn/T/ipykernel_22801/4088234538.py:18: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  movies_metadata_df = pd.read_csv(os.path.join(extract_folder, "movies_metadata.csv"))


credits_df shape: (45476, 3)
keywords_df shape: (46419, 2)
links_small_df shape: (9125, 3)
links_df shape: (45843, 3)
movies_metadata_df shape: (45466, 24)
ratings_small_df shape: (100004, 4)
ratings_df shape: (26024289, 4)
   adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   



# **Initial Insights From the Head of movies_metadata_df**

# Example Row

Toy Story (the first entry) has budget=30000000, revenue=373554033.0, runtime=81.0, status='Released', vote_average=7.7, and vote_count=5415.0. That’s a classic success story: relatively modest budget compared to a high box-office revenue.

---

# JSON Fields

belongs_to_collection has an ID and name referencing “Toy Story Collection.”
genres includes a list of genre objects, like {'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, ....
spoken_languages might be [{'iso_639_1': 'en', 'name': 'English'}].

---

# Variation in Data

Some movies show a budget of zero, indicating missing or unrecorded data.
Some columns are text-based (like overview, tagline), which you can use for content-based recommendation (e.g., using NLP techniques).


In [4]:
dataframes = {
    'credits_df': credits_df,
    'keywords_df': keywords_df,
    'links_small_df': links_small_df,
    'links_df': links_df,
    'movies_metadata_df': movies_metadata_df,
    'ratings_small_df': ratings_small_df,
    'ratings_df': ratings_df
}

for name, df in dataframes.items():
    print(f"\n=== {name} ===")
    print("Shape:", df.shape)
    
    # .info() summarizes dtypes, non-null counts, etc. good to know what we are working with for model compatibility
    print(df.info())
    

    print(df.head(10))



=== credits_df ===
Shape: (45476, 3)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45476 entries, 0 to 45475
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   cast    45476 non-null  object
 1   crew    45476 non-null  object
 2   id      45476 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 1.0+ MB
None
                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   
5  [{'cast_id': 25, 'character': 'Lt. Vincent Han...   
6  [{'cast_id': 1, 'character': 'Linus Larrabee',...   
7  [{'cast_id': 2, 'character': 'Tom Sawyer', 'cr...   
8  [{'cast_id': 1, 'character': 'Darren Francis T...   
9  [{'cast_id': 1, 'character': 'James Bond', 'cr...   

  

# **THE DATA**

## 1. **credits_df**
- **Shape:** (45,476 rows, 3 columns)
- **Columns:**
  - `cast`: JSON-like strings detailing cast members
  - `crew`: JSON-like strings detailing crew members
  - `id`: Likely the movie’s TMDb ID
- **No Nulls:** Each column is fully populated.
- **Potential Use:**  
  - Content-based recommendation features (e.g., which actors/directors a user tends to like)
  - The `id` here presumably matches the `id` in `movies_metadata_df` (both have 45k-ish rows), which makes merging straightforward.

---

## 2. **keywords_df**
- **Shape:** (46,419 rows, 2 columns)
- **Columns:**
  - `id`: Likely the TMDb ID again
  - `keywords`: JSON-like array of “tags” describing the movie
- **No Nulls:** Both columns fully populated.
- **Potential Use:**  
  - Another rich text-based or categorical feature source for content-based filtering or similarity searches.
  - Merge on `id` with `movies_metadata_df`.

---

## 3. **links_small_df**
- **Shape:** (9,125 rows, 3 columns)
- **Columns:** `movieId`, `imdbId`, `tmdbId`
  - `movieId` is the MovieLens ID
  - `imdbId` is the IMDb ID
  - `tmdbId` is the TMDb ID
- **Notes:**
  - `tmdbId` has 9,112 non-null values out of 9,125 (most are present, a few missing).
  - Smaller subset of movies (9,125) in the MovieLens rating system, presumably used in the “ml-latest-small” dataset.
- **Potential Use:**  
  - Mapping the MovieLens IDs (used in `ratings_small_df`) to TMDb IDs (used in `movies_metadata_df`). If you want to do a user-based or item-based collaborative filter using the smaller ratings set, you can join to `movies_metadata_df` or the other TMDb-based files.

---

## 4. **links_df**
- **Shape:** (45,843 rows, 3 columns)
- **Same columns** as `links_small_df` but for a larger set of MovieLens movie IDs.
- **tmdbId** missing for (45,843 - 45,624) = 219 movies.
- **Potential Use:**  
  - This is the “full” version if you plan on using the massive `ratings_df` (26 million ratings). 
  - A direct map from MovieLens IDs to TMDb IDs that should more closely match the 45k+ movies in `movies_metadata_df`.

---

## 5. **movies_metadata_df**
- **Shape:** (45,466 rows, 24 columns)
- **Columns** (high-level):
  - **Identifiers / Booleans:** `adult`, `id`, `imdb_id`, possibly a mismatch in data types (object vs. int/float).
  - **Descriptive / Text Fields:** `title`, `original_title`, `overview`, `tagline`, `homepage`.
  - **Structured Fields (Often JSON-like):** `belongs_to_collection`, `genres`, `spoken_languages`, `production_companies`, etc.
  - **Numeric-ish Fields:** `budget` (currently object, might need conversion), `revenue` (float), `vote_average` (float), `vote_count` (float).
  - **Dates / Status:** `release_date`, `status`.
- **Nulls / Missing Values:**  
  - Some columns (like `belongs_to_collection`, `homepage`, `tagline`) have a lot of NaNs.  
  - `budget` has a lot of zeros or is stored as strings—will need cleaning.
- **Potential Use:**  
  - **Primary “content” reference** for each movie, including textual and numeric attributes.  
  - Likely your main table to merge with `credits_df` & `keywords_df` (on TMDb `id`) and optionally join to rating data (through links).

---

## 6. **ratings_small_df**
- **Shape:** (100,004 rows, 4 columns)
- **Columns:** `userId`, `movieId`, `rating`, `timestamp`
- **No Nulls:** Data is complete for those 100k interactions.
- **Potential Use:**  
  - A **lightweight** dataset from MovieLens if you want to prototype quickly.  
  - Since `movieId` belongs to the MovieLens system, you’d use either `links_small_df` or `links_df` to map those IDs to TMDb, if you want to combine with the metadata.

---

## 7. **ratings_df**
- **Shape:** (26,024,289 rows, 4 columns)
- **Columns:** `userId`, `movieId`, `rating`, `timestamp`
- **Potential Use:**  
  - The **full** rating dataset from MovieLens.  
  - Substantial coverage: 26+ million user–movie interactions. This can power a robust collaborative filtering model.  
  - Merging logic is the same as with `ratings_small_df`, but watch out for memory constraints and performance. 
  - You’d typically link to `links_df` to get the matching TMDb ID, then join with the main metadata.

---

## Summary & Next Steps

1. **Data Structure**:  
   - We have **TMDb-based** tables (`movies_metadata_df`, `credits_df`, `keywords_df`) all referencing a shared `id`.  
   - We have **MovieLens-based** rating data (`ratings_df` or `ratings_small_df`) that references `movieId`.  
   - We have **link** tables (`links_df` or `links_small_df`) to map between the MovieLens and TMDb ID systems.

2. **Which to Merge?**  
   - Since we plan on building a content-based recommender, we'll merge `movies_metadata_df`, `credits_df`, and `keywords_df` on their TMDb IDs (the `id` column). This will provide a rich set of features (genres, cast, crew, keywords).  


## Merge movies_metadata_df, credits_df, and keywords_df on their shared TMDb ID (id in each file).

In [5]:
import pandas as pd
import numpy as np

# 1. Convert 'id' in movies_metadata_df from object to numeric.
#    We use errors='coerce' so that any invalid strings become NaN.
movies_metadata_df['id'] = pd.to_numeric(movies_metadata_df['id'], errors='coerce')

# 2. Drop rows where 'id' is NaN (since we can’t merge on missing IDs).
movies_metadata_df = movies_metadata_df.dropna(subset=['id'])

# 3. Convert 'id' to integer so it lines up with credits_df and keywords_df (which have int64).
movies_metadata_df['id'] = movies_metadata_df['id'].astype(int)

# 4. Merge movies_metadata_df with credits_df on 'id' (TMDb ID).
#    Left join ensures we keep all rows from movies_metadata_df, 
#    even if some movies don't appear in credits_df.
merged_df = movies_metadata_df.merge(credits_df, how='left', on='id')

# 5. Merge the result with keywords_df on 'id'.
merged_df = merged_df.merge(keywords_df, how='left', on='id')

# 6. Inspect the final merged DataFrame
print("Merged DataFrame shape:", merged_df.shape)
print(merged_df.info())
print(merged_df.head())


Merged DataFrame shape: (46629, 27)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46629 entries, 0 to 46628
Data columns (total 27 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  46629 non-null  object 
 1   belongs_to_collection  4574 non-null   object 
 2   budget                 46629 non-null  object 
 3   genres                 46629 non-null  object 
 4   homepage               8009 non-null   object 
 5   id                     46629 non-null  int64  
 6   imdb_id                46612 non-null  object 
 7   original_language      46618 non-null  object 
 8   original_title         46629 non-null  object 
 9   overview               45634 non-null  object 
 10  popularity             46625 non-null  object 
 11  poster_path            46230 non-null  object 
 12  production_companies   46625 non-null  object 
 13  production_countries   46625 non-null  object 
 14  release_date      

/var/folders/3y/w5ts8qfs3sl47n2k1dprbgs00000gn/T/ipykernel_22801/1421010709.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies_metadata_df['id'] = movies_metadata_df['id'].astype(int)


1. Some “id” values appear multiple times in either credits_df or keywords_df. When pandas merged one row from movies_metadata_df with multiple matching rows from the other DataFrame, i ended up duplicating that movie’s row for each match.

2. Some “id” values didn’t exist in movies_metadata_df but appeared in credits_df or keywords_df. In a left merge, they won’t add brand new rows by themselves, but if there’s any mismatch in how IDs are interpreted, the result can get tricky.

---
## **1. Duplicate IDs**
   
a) Confirm movies_metadata_df’s ID Uniqueness

In [6]:
# How many unique IDs did we have before merging?
print("Unique IDs in movies_metadata_df before merge:",
    movies_metadata_df['id'].nunique())

# Compare to number of rows
print("Rows in movies_metadata_df:", len(movies_metadata_df))

Unique IDs in movies_metadata_df before merge: 45433
Rows in movies_metadata_df: 45463


b) Investigate Post-Merge Duplicates

In [7]:
print("Unique IDs in merged_df:", merged_df['id'].nunique())
print("Rows in merged_df:", len(merged_df))

Unique IDs in merged_df: 45433
Rows in merged_df: 46629


c) Inspect an Example Duplicate

In [8]:
dup_ids = merged_df[merged_df.duplicated(subset=['id'], keep=False)]
dup_ids.sort_values(by='id').head(10)

adult belongs_to_collection budget  \
37082  False                   NaN      0   
37083  False                   NaN      0   
36536  False                   NaN      0   
36537  False                   NaN      0   
26038  False                   NaN      0   
26039  False                   NaN      0   
36805  False                   NaN      0   
36804  False                   NaN      0   
36150  False                   NaN      0   
36151  False                   NaN      0   

                                                  genres homepage    id  \
37082                      [{'id': 18, 'name': 'Drama'}]      NaN  1998   
37083                      [{'id': 18, 'name': 'Drama'}]      NaN  1998   
36536  [{'id': 27, 'name': 'Horror'}, {'id': 878, 'na...      NaN  3025   
36537  [{'id': 27, 'name': 'Horror'}, {'id': 878, 'na...      NaN  3025   
26038  [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...      NaN  3057   
26039  [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...      NaN  3057   
36805  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...      NaN  3692   
36804  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...      NaN  3692   
36150                   [{'id': 53, 'name': 'Thriller'}]      NaN  4459   
36151                   [{'id': 53, 'name': 'Thriller'}]      NaN  4459   

         imdb_id original_language               original_title  \
37082  tt0384177                en                    Holy Lola   
37083  tt0384177                en                    Holy Lola   
36536  tt0054416                en  The Two Faces of Dr. Jekyll   
36537  tt0054416                en  The Two Faces of Dr. Jekyll   
26038  tt0368730                en                 Frankenstein   
26039  tt0368730                en                 Frankenstein   
36805  tt0060956                en   New York chiama Superdrago   
36804  tt0060956                en   New York chiama Superdrago   
36150  tt0044967                en          Night Without Sleep   
36151  tt0044967                en          Night Without Sleep   

                                                overview  ...  \
37082  The story of a young couple, Pierre and Gerald...  ...   
37083  The story of a young couple, Pierre and Gerald...  ...   
36536  After a series of scientific experiments direc...  ...   
36537  After a series of scientific experiments direc...  ...   
26038  Frankenstein is a 2004 U.S. television miniser...  ...   
26039  Frankenstein is a 2004 U.S. television miniser...  ...   
36805  A series of murders in Michigan lead an Americ...  ...   
36804  A series of murders in Michigan lead an Americ...  ...   
36150  Awaking one morning out of a drunken stupor, c...  ...   
36151  Awaking one morning out of a drunken stupor, c...  ...   

                                        spoken_languages    status  \
37082  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
37083  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
36536           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
36537           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
26038          [{'iso_639_1': 'fr', 'name': 'Français'}]  Released   
26039          [{'iso_639_1': 'fr', 'name': 'Français'}]  Released   
36805          [{'iso_639_1': 'it', 'name': 'Italiano'}]  Released   
36804          [{'iso_639_1': 'it', 'name': 'Italiano'}]  Released   
36150           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
36151           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                                 tagline  \
37082                                                NaN   
37083                                                NaN   
36536                   LIKE NOTHING YOU HAVE EVER SEEN!   
36537                   LIKE NOTHING YOU HAVE EVER SEEN!   
26038                                                NaN   
26039                                                NaN   
36805  Beautiful Women! Silver-Masked Me

**Drop Exact Duplicates (If Data is Identical)**

In [10]:
credits_df = credits_df.drop_duplicates(subset=['id', 'cast', 'crew'])
keywords_df = keywords_df.drop_duplicates(subset=['id', 'keywords'])

In [11]:
merged_df.drop_duplicates(inplace=True)

In [12]:
# Check final row count and uniqueness
print(len(merged_df))
print(merged_df['id'].nunique())


45463
45433


In [13]:
# Identify IDs that occur more than once
dup_id_counts = (
    movies_metadata_df.groupby('id')
    .size()
    .reset_index(name='count')
    .query('count > 1')
)

print(f"Number of IDs with duplicates: {len(dup_id_counts)}")
dup_id_counts.head(10)


Number of IDs with duplicates: 29


id  count
2129    4912      2
2295    5511      2
4563   10991      2
4667   11115      2
5585   12600      2
5854   13209      2
6781   14788      2
6912   15028      2
8366   18440      2
10006  22649      2

In [16]:
dup_id = 10991
movies_metadata_df[movies_metadata_df['id'] == dup_id]

adult                              belongs_to_collection    budget  \
4114   False  {'id': 34055, 'name': 'Pokémon Collection', 'p...  16000000   
44821  False  {'id': 34055, 'name': 'Pokémon Collection', 'p...  16000000   

                                                  genres  \
4114   [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
44821  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

                                homepage     id    imdb_id original_language  \
4114   http://movies.warnerbros.com/pk3/  10991  tt0235679                ja   
44821  http://movies.warnerbros.com/pk3/  10991  tt0235679                ja   

             original_title  \
4114   Pokémon 3: The Movie   
44821  Pokémon 3: The Movie   

                                                overview  ... release_date  \
4114   When Molly Hale's sadness of her father's disa...  ...   2000-07-08   
44821  When Molly Hale's sadness of her father's disa...  ...   2000-07-08   

          revenue runtime                          spoken_languages    status  \
4114   68411275.0    93.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   
44821  68411275.0    93.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                             tagline                          title  video  \
4114   Pokémon: Spell of the Unknown  Pokémon: Spell of the Unknown  False   
44821  Pokémon: Spell of the Unknown  Pokémon: Spell of the Unknown  False   

      vote_average vote_count  
4114           6.0      143.0  
44821          6.0      144.0  

[2 rows x 24 columns]

In [17]:
# Drop duplicate rows that share the same 'id', keeping only the first occurrence
movies_metadata_df = movies_metadata_df.drop_duplicates(subset=['id'], keep='first')

# Verify that rows now match the count of unique IDs
print("After dropping duplicates:")
print("Rows in movies_metadata_df:", len(movies_metadata_df))
print("Unique IDs:", movies_metadata_df['id'].nunique())

After dropping duplicates:
Rows in movies_metadata_df: 45433
Unique IDs: 45433


**Re-Merge DF**

In [18]:
merged_df = movies_metadata_df.merge(credits_df, how='left', on='id')
merged_df = merged_df.merge(keywords_df, how='left', on='id')

In [19]:
print("Merged DataFrame rows:", len(merged_df))
print("Merged DataFrame unique IDs:", merged_df['id'].nunique())

Merged DataFrame rows: 45440
Merged DataFrame unique IDs: 45433


In [21]:
dupes = (
    merged_df[merged_df.duplicated(subset=['id'], keep=False)]
    .sort_values(by='id')
)

print("Number of rows involved in duplicates:", len(dupes))
dupes.head(10)


Number of rows involved in duplicates: 14


adult belongs_to_collection    budget  \
5866   False                   NaN  30000000   
5867   False                   NaN  30000000   
9576   False                   NaN  10000000   
9577   False                   NaN  10000000   
16165  False                   NaN         0   
16166  False                   NaN         0   
4356   False                   NaN   3512454   
4357   False                   NaN   3512454   
838    False                   NaN         0   
839    False                   NaN         0   

                                                  genres homepage      id  \
5866   [{'id': 35, 'name': 'Comedy'}, {'id': 80, 'nam...      NaN    4912   
5867   [{'id': 35, 'name': 'Comedy'}, {'id': 80, 'nam...      NaN    4912   
9576   [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...      NaN   69234   
9577   [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...      NaN   69234   
16165  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...      NaN   99080   
16166  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...      NaN   99080   
4356                       [{'id': 18, 'name': 'Drama'}]      NaN  110428   
4357                       [{'id': 18, 'name': 'Drama'}]      NaN  110428   
838                        [{'id': 18, 'name': 'Drama'}]      NaN  132641   
839                        [{'id': 18, 'name': 'Drama'}]      NaN  132641   

         imdb_id original_language                   original_title  \
5866   tt0270288                en  Confessions of a Dangerous Mind   
5867   tt0270288                en  Confessions of a Dangerous Mind   
9576   tt0100361                en         The Phantom of the Opera   
9577   tt0100361                en         The Phantom of the Opera   
16165  tt0022537                en                       The Viking   
16166  tt0022537                en                       The Viking   
4356   tt2018086                fr             Camille Claudel 1915   
4357   tt2018086                fr             Camille Claudel 1915   
838    tt0046468                ja                            Tsuma   
839    tt0046468                ja                            Tsuma   

                                                overview  ...  \
5866   Television made him famous, but his biggest hi...  ...   
5867   Television made him famous, but his biggest hi...  ...   
9576   Count de Chagnie has discovered Christine's si...  ...   
9577   Count de Chagnie has discovered Christine's si...  ...   
16165  Originally called White Thunder, American prod...  ...   
16166  Originally called White Thunder, American prod...  ...   
4356   Winter, 1915. Confined by her family to an asy...  ...   
4357   Winter, 1915. Confined by her family to an asy...  ...   
838    Ten years into a marriage, the wife is disappo...  ...   
839    Ten years into a marriage, the wife is disappo...  ...   

                                        spoken_languages    status  \
5866            [{'iso_639_1': 'en', 'name': 'English'}]  Released   
5867            [{'iso_639_1': 'en', 'name': 'English'}]  Released   
9576   [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
9577   [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
16165           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
16166           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
4356           [{'iso_639_1': 'fr', 'name': 'Français'}]  Released   
4357           [{'iso_639_1': 'fr', 'name': 'Français'}]  Released   
838                 [{'iso_639_1': 'ja', 'name': '日本語'}]  Released   
839                 [{'iso_639_1': 'ja', 'name': '日本語'}]  Released   

                                                 tagline  \
5866             Some things are better left top secret.   
5867             Some things are better left top secret.   
9576                                                 NaN   
9577                                                 NaN   
16165  Actually produced during the Great Newfoundlan...   
1

In [23]:
# Rows I want to remove 
rows_to_drop = [5867, 9577, 16166, 4356, 838]

# Drop them in-place
merged_df.drop(index=rows_to_drop, inplace=True)

# Verify new row count vs. unique ID count
print("Rows in merged_df:", len(merged_df))
print("Unique IDs in merged_df:", merged_df['id'].nunique())


Rows in merged_df: 45435
Unique IDs in merged_df: 45433


In [24]:
dupes = (
    merged_df[merged_df.duplicated(subset=['id'], keep=False)]
    .sort_values(by='id')
)

print("Number of rows involved in duplicates:", len(dupes))
dupes.head(10)

Number of rows involved in duplicates: 4


adult                              belongs_to_collection  budget  \
13220  False  {'id': 158365, 'name': 'Why We Fight', 'poster...       0   
13221  False  {'id': 158365, 'name': 'Why We Fight', 'poster...       0   
33167  False                                                NaN  980000   
33168  False                                                NaN  980000   

                                                  genres homepage      id  \
13220                [{'id': 99, 'name': 'Documentary'}]      NaN  159849   
13221                [{'id': 99, 'name': 'Documentary'}]      NaN  159849   
33167  [{'id': 18, 'name': 'Drama'}, {'id': 14, 'name...      NaN  298721   
33168  [{'id': 18, 'name': 'Drama'}, {'id': 14, 'name...      NaN  298721   

         imdb_id original_language                    original_title  \
13220  tt0173769                en  Why We Fight: Divide and Conquer   
13221  tt0173769                en  Why We Fight: Divide and Conquer   
33167  tt2818654                th                     รักที่ขอนแก่น   
33168  tt2818654                th                     รักที่ขอนแก่น   

                                                overview  ...  \
13220  The third film of Frank Capra's 'Why We Fight"...  ...   
13221  The third film of Frank Capra's 'Why We Fight"...  ...   
33167  In a hospital, ten soldiers are being treated ...  ...   
33168  In a hospital, ten soldiers are being treated ...  ...   

                                        spoken_languages    status tagline  \
13220           [{'iso_639_1': 'en', 'name': 'English'}]  Released     NaN   
13221           [{'iso_639_1': 'en', 'name': 'English'}]  Released     NaN   
33167  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released     NaN   
33168  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released     NaN   

                                  title  video  vote_average  vote_count  \
13220  Why We Fight: Divide and Conquer  False           5.0         1.0   
13221  Why We Fight: Divide and Conquer  False           5.0         1.0   
33167             Cemetery of Splendour  False           4.4        50.0   
33168             Cemetery of Splendour  False           4.4        50.0   

                                                    cast  \
13220  [{'cast_id': 12, 'character': 'Narrator', 'cre...   
13221  [{'cast_id': 12, 'character': 'Narrator', 'cre...   
33167  [{'cast_id': 1, 'character': 'Jenjira', 'credi...   
33168  [{'cast_id': 1, 'character': 'Jenjira', 'credi...   

                                                    crew  \
13220  [{'credit_id': '52fe4c229251416c910f10d5', 'de...   
13221  [{'credit_id': '52fe4c229251416c910f10d5', 'de...   
33167  [{'credit_id': '5448c8efc3a3680fb4001582', 'de...   
33168  [{'credit_id': '5739ceeac3a3683c9d000c12', 'de...   

                                                keywords  
13220                                                 []  
13221                                                 []  
33167  [{'id': 1227, 'name': 'cemetery'}, {'id': 1261...  
33168  [{'id': 1227, 'name': 'cemetery'}, {'id': 1261...  

[4 rows x 27 columns]

In [25]:
rows_to_drop = [13221, 33168]

merged_df.drop(index=rows_to_drop, inplace=True)

print("Rows in merged_df:", len(merged_df))
print("Unique IDs in merged_df:", merged_df['id'].nunique())

Rows in merged_df: 45433
Unique IDs in merged_df: 45433


In [26]:
movies_df = merged_df.copy()

# **Duplicates Complete**

---

In [31]:
import numpy as np

# Any row with budget exactly 8,000,000.0 is reverted to NaN
movies_df.loc[movies_df['budget'] == 8000000.0, 'budget'] = np.nan

# Confirm the result
print(movies_df['budget'].head(10))
movies_df['budget'].isnull().sum()

0    30000000.0
1    65000000.0
2           NaN
3    16000000.0
4           NaN
5    60000000.0
6    58000000.0
7           NaN
8    35000000.0
9    58000000.0
Name: budget, dtype: float64


36708

**Keep them as NaN and let the model handle missing values ( potentially by ignoring that feature or using an imputation strategy at model time).**

In [32]:
import numpy as np

# Convert revenue
movies_df['revenue'] = pd.to_numeric(movies_df['revenue'], errors='coerce')
movies_df.loc[movies_df['revenue'] == 0, 'revenue'] = np.nan

# Convert popularity
movies_df['popularity'] = pd.to_numeric(movies_df['popularity'], errors='coerce')
movies_df.loc[movies_df['popularity'] == 0, 'popularity'] = np.nan

# Convert vote_average, vote_count
movies_df['vote_average'] = pd.to_numeric(movies_df['vote_average'], errors='coerce')
movies_df['vote_count'] = pd.to_numeric(movies_df['vote_count'], errors='coerce')

# Convert runtime
movies_df['runtime'] = pd.to_numeric(movies_df['runtime'], errors='coerce')
movies_df.loc[movies_df['runtime'] == 0, 'runtime'] = np.nan

# Quick verification
print(movies_df[['revenue', 'popularity', 'vote_average', 'vote_count', 'runtime']].info())
print(movies_df[['revenue', 'popularity', 'vote_average', 'vote_count', 'runtime']].head())


<class 'pandas.core.frame.DataFrame'>
Index: 45433 entries, 0 to 45439
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   revenue       7398 non-null   float64
 1   popularity    45364 non-null  float64
 2   vote_average  45430 non-null  float64
 3   vote_count    45430 non-null  float64
 4   runtime       43615 non-null  float64
dtypes: float64(5)
memory usage: 2.1 MB
None
       revenue  popularity  vote_average  vote_count  runtime
0  373554033.0   21.946943           7.7      5415.0     81.0
1  262797249.0   17.015539           6.9      2413.0    104.0
2          NaN   11.712900           6.5        92.0    101.0
3   81452156.0    3.859495           6.1        34.0    127.0
4   76578911.0    8.387519           5.7       173.0    106.0


**Data Types (Dtype)**:

1. All five columns are now float64, which is exactly what you want if you plan on doing numeric analyses or feeding them into a model.

**Non-Null Counts:**

1. revenue has 7,398 non-null entries out of 45,433 total rows. This means the majority—nearly 38,000—are now NaN. That often happens because real-world movie metadata frequently lacks valid revenue data, especially for older or more obscure titles.

2. popularity still has a high fill rate (45,364), so not much was missing or zero there.

3. vote_average and vote_count are nearly complete, which makes sense—TMDb usually has user ratings for most titles.

4. runtime has 43,615 non-null entries, meaning about 1,800 are missing or were zero and got turned into NaN.

**Sample Values:**

1. Row 2 has revenue = NaN, reflecting that it was originally 0 or invalid, now set to NaN.

 Having so many NaNs in revenue might seem extreme, but it’s not unusual in open movie databases and won't be critical for the success of our model.

**2. Convert release_date to a Datetime**

In [33]:
movies_df['release_date'] = pd.to_datetime(movies_df['release_date'], errors='coerce')

In [34]:
print("Missing release dates:", movies_df['release_date'].isnull().sum())
print(movies_df[['title', 'release_date']].head())

Missing release dates: 87
                         title release_date
0                    Toy Story   1995-10-30
1                      Jumanji   1995-12-15
2             Grumpier Old Men   1995-12-22
3            Waiting to Exhale   1995-12-22
4  Father of the Bride Part II   1995-02-10


In [35]:
movies_df.dropna(subset=['release_date'], inplace=True)


**JSON Parsing**

In [36]:
import ast

def parse_json_column(df, column):
    def try_parse(val):
        if isinstance(val, str):
            try:
                return ast.literal_eval(val)
            except:
                return []
        elif isinstance(val, list):
            return val
        return []
    
    df[column] = df[column].apply(try_parse)

json_cols = ['genres', 'keywords', 'cast', 'crew', 'spoken_languages',
            'production_companies', 'production_countries', 'belongs_to_collection']

for col in json_cols:
    if col in movies_df.columns:
        parse_json_column(movies_df, col)

**JSON parsing is crucial because it transforms textual representations of structured data into actual Python structures. 
This makes it far easier to do the content-based part of “content-based recommendations,”**

In [37]:
import random

rand_index = random.choice(movies_df.index)
cell_value = movies_df.loc[rand_index, 'genres']
print(f"Random index: {rand_index}")
print(cell_value)
print("Type:", type(cell_value))


Random index: 20836
[{'id': 28, 'name': 'Action'}, {'id': 53, 'name': 'Thriller'}]
Type: <class 'list'>


In [39]:
movies_df.info()

movies_df.isnull().sum().sort_values(ascending=False)

movies_df.head(5)

<class 'pandas.core.frame.DataFrame'>
Index: 45346 entries, 0 to 45439
Data columns (total 27 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   adult                  45346 non-null  object        
 1   belongs_to_collection  45346 non-null  object        
 2   budget                 8721 non-null   float64       
 3   genres                 45346 non-null  object        
 4   homepage               7761 non-null   object        
 5   id                     45346 non-null  int64         
 6   imdb_id                45332 non-null  object        
 7   original_language      45335 non-null  object        
 8   original_title         45346 non-null  object        
 9   overview               44405 non-null  object        
 10  popularity             45306 non-null  float64       
 11  poster_path            45007 non-null  object        
 12  production_companies   45346 non-null  object        
 13  produc

adult                              belongs_to_collection      budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000.0   
1  False                                                 []  65000000.0   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         NaN   
3  False                                                 []  16000000.0   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         NaN   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ...  \
0  Led by Woody, Andy's toys live happily in his ...  ...   
1  When siblings Judy and Peter discover an encha...  ...   
2  A family wedding reignites the ancient feud be...  ...   
3  Cheated on, mistreated and stepped on, the wom...  ...   
4  Just when George Banks has recovered from his ...  ...   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
3           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
4           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                             tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   
3  Friends are the people who let you be yourself...   
4  Just When His World Is Back To Normal... He's ...   

                         title  video  vote_average  vote_count  \
0                    Toy Story  False           7.7      5415.0   
1                      Jumanji  False           6.9      2413.0   
2             Grumpier Old Men  False           6.5        92.0   
3            Waiting to Exhale  False           6.1        34.0   
4  Father of the Bride Part II  False           5.7       173.0   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': 'Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...   
3  [{'credit_id': '52fe44779251416c91011acb', 'de...   
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...   

                                            keywords  
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...  
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...  
2  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...  
3  [{'id': 818, 'name': 'based on novel'}, {'id':...  
4  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...  

[5 rows x 27 columns]

**Only about 19% of rows have a non-null budget and 16% have revenue. That’s a lot of missing data**

**Strategy: Missing Indicator + Impute**

A popular approach for neural nets is to add a binary flag for missingness.

This way, the neural net can learn that “budget was originally missing” if that’s a relevant pattern.

In [40]:
movies_df['budget_missing'] = movies_df['budget'].isnull().astype(int) #This will be 1 for rows where budget was originally missing, and 0 otherwise.

median_budget = movies_df['budget'].median()  # This will Impute the median budget for missing values
movies_df['budget'].fillna(median_budget, inplace=True)   # This will fill the missing values with the median budget



# Now the model has two pieces of information for each movie’s budget:
# the budget value itself (which is the median if originally missing).
# A flag (budget_missing) indicating if the original was NaN or not.

/var/folders/3y/w5ts8qfs3sl47n2k1dprbgs00000gn/T/ipykernel_22801/4043059309.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  movies_df['budget'].fillna(median_budget, inplace=True)


In [41]:
print("Missing budgets:", movies_df['budget'].isnull().sum())
print("Sample budgets:\n", movies_df['budget'].head(10))

Missing budgets: 0
Sample budgets:
 0    30000000.0
1    65000000.0
2     9000000.0
3    16000000.0
4     9000000.0
5    60000000.0
6    58000000.0
7     9000000.0
8    35000000.0
9    58000000.0
Name: budget, dtype: float64


In [42]:
import ast

# ----- LET'S FINISH THIS:
# STEP 1: Impute 'revenue' & 'runtime' + Create Missing Indicators -----

# 1A. Revenue
movies_df['revenue_missing'] = movies_df['revenue'].isnull().astype(int)
median_revenue = movies_df['revenue'].median()
movies_df['revenue'] = movies_df['revenue'].fillna(median_revenue)

# 1B. Runtime
movies_df['runtime_missing'] = movies_df['runtime'].isnull().astype(int)
median_runtime = movies_df['runtime'].median()
movies_df['runtime'] = movies_df['runtime'].fillna(median_runtime)

# ----- STEP 2: Drop Low-Value or Mostly-Missing Columns -----
# 'homepage' (~17% coverage), 'tagline' (~45%), 'adult' & 'video' are rarely useful for a content-based recommender.
columns_to_drop = ['homepage', 'tagline', 'adult', 'video']
movies_df.drop(columns=columns_to_drop, inplace=True, errors='ignore')

# ----- STEP 3: Keep Only "Released" Movies & Drop 'status' -----
movies_df = movies_df[movies_df['status'] == "Released"]
movies_df.drop(columns=['status'], inplace=True, errors='ignore')

# ----- STEP 4: Ensure JSON-Like Columns Are Parsed -----
# This code will parse these columns so each row is a Python list/dict (not a string).
json_cols = [
    'genres', 'keywords', 'cast', 'crew', 
    'spoken_languages', 'production_companies',
    'production_countries', 'belongs_to_collection'
]

def parse_json_column(df, col):
    def try_parse(val):
        if isinstance(val, str):
            try:
                return ast.literal_eval(val)
            except:
                return []
        elif isinstance(val, list):
            return val
        return []
    df[col] = df[col].apply(try_parse)

for col in json_cols:
    if col in movies_df.columns:
        parse_json_column(movies_df, col)

# ----- STEP 5: Final Check -----
print("After final cleaning:")
print(movies_df.info())
print(movies_df.isnull().sum().sort_values(ascending=False).head(10))

# Itty-bitty peek at a few rows
print(movies_df.head(3))


After final cleaning:
<class 'pandas.core.frame.DataFrame'>
Index: 44907 entries, 0 to 45439
Data columns (total 25 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   belongs_to_collection  44907 non-null  object        
 1   budget                 44907 non-null  float64       
 2   genres                 44907 non-null  object        
 3   id                     44907 non-null  int64         
 4   imdb_id                44895 non-null  object        
 5   original_language      44897 non-null  object        
 6   original_title         44907 non-null  object        
 7   overview               44000 non-null  object        
 8   popularity             44869 non-null  float64       
 9   poster_path            44575 non-null  object        
 10  production_companies   44907 non-null  object        
 11  production_countries   44907 non-null  object        
 12  release_date           44907 non-null  date

---

# **CLEANING FINISHED**


---

### Final Cleaning Steps Overview

1. **Dropped Duplicates** in `movies_metadata_df` to ensure each TMDb `id` corresponds to a single row.  
   
2. **Merged** Movie DataFrames  
   - Combined `movies_metadata`, `credits`, and `keywords` on the shared TMDb ID, ensuring each movie’s metadata, cast/crew, and keywords ended up in a single DataFrame.  
   - Resolved any cross-join duplicates (e.g., rows that appeared multiple times for the same ID) by dropping or unifying them.  
  
3. **Converted Numeric-Like Columns**  
   - `budget`, `revenue`, `popularity`, `vote_average`, `vote_count`, and `runtime` from strings (object dtype) to `float64`, turning zeros into `NaN` when appropriate. 
   
4. **Dropped or Imputed Missing Values** for Key Columns  
   - **Budget, Revenue, Runtime**: Created missingness indicator columns (`budget_missing`, `revenue_missing`, `runtime_missing`) and filled the missing values with the median for each column.  
   - Ensured that the model can see which movies had originally missing/unknown values.  
  
5. **Dropped Rows with Missing or Invalid Release Dates**  
   - Ensured that `release_date` was converted to a proper datetime, then dropped the small fraction of rows lacking valid dates.  
  
6. **Filtered Out Non-“Released” Movies**  
   - Kept only entries where `status` == “Released,” then dropped the `status` column.  
  
7. **Dropped Low-Value or Mostly-Empty Columns**  
   - Removed columns such as `homepage`, `tagline`, `adult`, and `video` that were often empty or added little relevance for our content-based recommender.  
  
8. **JSON Parsing** of List/Dict Columns  
   - Parsed columns like `genres`, `keywords`, `cast`, `crew`, etc., from JSON-like strings into actual Python lists/dicts, this provides easy extraction of genre names, keyword tokens, cast member data, and so on.  

9.  **Final Integrity Checks**  
   - Verified the DataFrame shape, types, and non-null counts with `.info()`, `.isnull().sum()`, and previewed data with `.head()` to confirm a clean, consistent dataset.
